Lagom: Replace LibJSGCVerifier with the more general ClangPlugin + Add tests!

mattco98 opened the pull request:
This is a pretty substantial rewrite, motivated by three factors:

- Directly sharing information between the different C++ threads, which means getting rid of the python wrapper script
- Better integration with Lagom, (i.e. invocation through `./Meta/serenity.sh`)
- Tests

Some demo screenshots:

<details><summary>Automatically capturing all JS-relevant files with `--scan js` (no more ugly `-p`!)</summary>

![image](https://github.com/SerenityOS/serenity/assets/28801118/cadae267-f884-4d1d-af72-45d4dd42f22b)

</details> 

<details><summary>Passing all of the current tests</summary>
<p>

![image](https://github.com/SerenityOS/serenity/assets/28801118/d8bc4e1f-5c4f-4419-ae87-03a555817c38)

</p>
</details> 

<details><summary>Failing a test that I intentionally messed up</summary>
<p>

![image](https://github.com/SerenityOS/serenity/assets/28801118/31403d4b-c66e-43b7-9f10-28653d65cfdf)

</p>
</details> 

A few loose ends:
- Bikeshedding the name, since it leaves a little to be desired at the moment
- My CMake skills are quite lacking, so I'm not quite sure how best to set it up, especially with the whole Lagom/host Clang toolchain thing. Should tests run with Lagom toolchain?
- I'm really not sure why this new configuration caused all of these `#include` errors to show up, but they're all valid errors, so I didn't look into it too hard. 

Some non-analysis future work:
- Don't return an error code if the only diagnostics we get are from Clang and not us (like not being able to resolve `<GL/gl.h>` or whatever), which will allow us to run this on CI
- Reduce the amount of duplicate work, which should be very easy now that all of the C++ threads can talk to each other.

mattco98 commented:
Yeah I'm definitely going to need help with the CI here. I tried both with and without `--lagom`, though that option only changes how the tool does internal processing; it doesn't actually affect the CMake config. I think CI needs to `LAGOM_BUILD` to `ON`? Though I'm not sure. It'd be great if the `--lagom` option could change that, but I'm not sure if that's possible.

awesomekling commented:
cc @ADKaster for build system advice :)

awesomekling commented:
Can this be integrated with CLion too? :)

mattco98 commented:
> Can this be integrated with CLion too? :)

I did some poking around and it seems like it would be a challenge, but it should be much easier to integrate it into the build. I think I'd just need to split it up into a lib and an executable, and then add the lib to the command line invocation.

ADKaster commented on Tests/ClangPlugin/test-runner.cpp:
```diff
@@ -0,0 +1,120 @@
+    return MUST(file->read_until_eof());
+}
+
+void find_tests(StringView path_to_search)
```
Going to take a look at the CMake here in a sec, but my question about this test runner is: Would it be useful to use lit and FileCheck for these tests instead of our own custom test runner? If our analysis is generic enough that we would want to upstream it into `clang-analyzer` or such, it would be useful to be using the upstream llvm-project test runner.

mattco98 replied:
Not sure about lit and FileCheck, I can look into those, but I don't think anything here is really upstreamable. Maybe some checks in the future will be (since this is now a more generic tool), but atm I don't think so.

mattco98 replied:
Rewrote the tests with `lit`

Hendiadyoin1 commented:
Also quick question, can we get clangd to load and use the plugin, if so maybe document how to do so

mattco98 commented:
> Also quick question, can we get clangd to load and use the plugin, if so maybe document how to do so

That's basically the same thing AK asked, and if it is I'd figure it out in a future PR, since this one is big enough as it is

mattco98 commented:
Split it up into separate shared libs and added it as a build command, which seems to work pretty well, but once again I'm not sure how to do the CMake side of it. It's not a dependency of anything, so it won't be built until you build it yourself, and even then you have to manually comment out the `-fplugin` lines since it'll try to add the plugin when compiling the plugin. I'm also wondering if, with this way, it'd be possible to just make two sets of libraries, one for host and one for serenity's toolchain. Otherwise I really don't know how to deal with this whole `LAGOM_BUILD` thing.

mattco98 commented:
Remaining issues:
- The dependencies seem to be hooked up, but for some reason the plugins are built in parallel to the thing they're required for? So the first two times the `test` is ran it fails, but afterwards it works. I probably don't understand either what an `ALL` target is, or the correct way to ensure that the clang plugins are a dependency of everything (right now I just set it as a dependency of AK, since I figured that is a dependency of everything else). This does not happen with `build lagom ...`.
- When I run `./Meta/serenity.sh build x86_64 Clang`, I get the following error:
  ```
  [0/1] Re-running CMake...
  -- Configuring done (0.0s)
  -- Generating done (0.0s)
  -- Build files have been written to: /home/matt/code/serenity/Build/superbuild-x86_64clang
  [0/6] Performing build step for 'lagom'
  [0/1] Re-running CMake...
  -- Configuring done (0.0s)
  -- Generating done (0.0s)
  -- Build files have been written to: /home/matt/code/serenity/Build/lagom
  ninja: no work to do.
  [1/6] Performing install step for 'lagom'
  [0/1] Install the project...
  -- Install configuration: ""
  CMake Error at cmake_install.cmake:46 (file):
    file cannot create directory: /usr/local/share/Lagom.  Maybe need
    administrative privileges.


  FAILED: CMakeFiles/install.util 
  cd /home/matt/code/serenity/Build/lagom && /usr/bin/cmake -P cmake_install.cmake
  ninja: build stopped: subcommand failed.
  FAILED: lagom-prefix/src/lagom-stamp/lagom-install /home/matt/code/serenity/Build/superbuild-x86_64clang/lagom-prefix/src/lagom-stamp/lagom-install 
  cd /home/matt/code/serenity/Build/lagom && /usr/bin/cmake --build . --target install
  ninja: build stopped: subcommand failed.
  ```

mattco98 commented:
At this point the PR description and reviews are outdated since the code is nothing like it was at the beginning, and I'm about to make another big change, so I'll close this PR and reopen another one when I'm ready.
